### PR TITLE
Constructors and abstract methods.

### PR DESCRIPTION
**Pull request by swankjesse**

### CODE REVIEW

File: src/main/java/com/squareup/javawriter/builders/CodeWriter.java [hunk 1]
```diff
@@ -55,13 +55,21 @@ public CodeWriter(StringBuilder out, ImmutableMap<ClassName, String> importedTyp
...
   }
 
   public CodeWriter unindent() {
-    checkState(indentLevel > 0);
-    indentLevel--;
+    return unindent(1);
+  }
+
+  public CodeWriter unindent(int levels) {
+    checkState(indentLevel - levels >= 0);
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javawriter/builders/CodeWriter.java]: Seems like it should be checkArgument w/message of some sort.

File: src/main/java/com/squareup/javawriter/builders/MethodSpec.java [hunk 2]
```diff
@@ -38,6 +40,11 @@
   public final ImmutableList<Snippet> snippets;
 
   private MethodSpec(Builder builder) {
+    checkArgument(builder.returnType != null ^ builder.name == Name.CONSTRUCTOR,
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javawriter/builders/MethodSpec.java]: xor. love it.

File: src/main/java/com/squareup/javawriter/builders/TypeSpec.java [hunk 3]
```diff
@@ -42,6 +43,17 @@
...
     checkArgument(builder.name != null ^ builder.anonymousTypeArguments != null,
         "types must have either a name or anonymous type arguments");
+    boolean isInterface = builder.type == Type.INTERFACE;
+    boolean typeIsAbstract = builder.modifiers.contains(Modifier.ABSTRACT) || isInterface;
+    for (MethodSpec methodSpec : builder.methodSpecs) {
+      checkArgument(typeIsAbstract || !methodSpec.hasModifier(Modifier.ABSTRACT),
+          "non-abstract type %s cannot declare abstract method %s", builder.name, methodSpec.name);
+      checkArgument(!isInterface || methodSpec.hasModifier(Modifier.ABSTRACT),
+          "interface %s cannot declare non-abstract method %s", builder.name, methodSpec.name);
+      checkArgument(!isInterface || methodSpec.hasModifier(Modifier.PUBLIC),
```
Comment (JakeWharton) [inline src/main/java/com/squareup/javawriter/builders/TypeSpec.java]: We'll have to go through and add tests which verify these conditions correctly fail.

Comment (JakeWharton) [inline src/main/java/com/squareup/javawriter/builders/TypeSpec.java]: Later, that is.

Comment (JakeWharton) [inline src/main/java/com/squareup/javawriter/builders/TypeSpec.java]: Need this `public` check for interface fields as well

Comment (swankjesse) [inline src/main/java/com/squareup/javawriter/builders/TypeSpec.java]: Fixed. Also static and final for fields.

**Comment by JakeWharton** [comment]
Sorry, LGTM. Was mesmerized by the awesome Retrofit service tests...

File: src/main/java/com/squareup/javawriter/builders/CodeWriter.java (hunk 1 above)
Comment (swankjesse) [inline src/main/java/com/squareup/javawriter/builders/CodeWriter.java]: Fixed. I don't think this needs a message; it's not a public interface.

**Pull request merged**